libparsec/crates/client/tests/unit/certificates_ops/store.rs

```diff
@@ -0,0 +1,190 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use std::sync::Arc;
+
+use libparsec_platform_storage::certificates::GetCertificateError;
+use libparsec_tests_fixtures::prelude::*;
+use libparsec_types::prelude::*;
+
+use super::utils::certificates_store_factory;
+
+#[parsec_test(testbed = "minimal", with_server)]
+async fn add_device_certificate(alice_device_certif: &DeviceCertificate, env: &TestbedEnv) {
```

Comment:
Regarding testing of the store: I think it's okay to have a couple of test on this internal component (given testing a small part of bigger system helps narrow down errors) just like you did here.

However don't try to go too far on this: `CertificateOps` is considered a public component ("public" meaning it is used by unrelated systems such as user/workspace ops and `libparsec/src`), so it api shouldn't change much and hence is a good fit for test

On the other hand `store.rs` is internal stuff and subject to change (guess what, I've already entirely rewritten it last week ^^)

So we want our test to use the more stable `CertificateOps` api so that we can do our refactoring with confidence.

Reply:
Aaaand I realize `add_certificates_batch` is not part of the `CertificateOps` api

In theory this is is not needed given only `poll_server_for_new_certificates` needs to call it.

But I wonder if we shouldn't expose it there anyway (as `pub(crate)`, maybe with a `#[test]`). The idea is to make it explicit this function is considered stable api and we heavily rely on it for our tests.

I'm not sure how useful it is though: once you've written a ton of tests that relies on this `certfificates_ops::add::add_certificates_batch` function, it becomes pretty obvious this function should not be modified

It can also be convenient to do `certificate_ops.add_certificates_batch(...)` in your tests instead of the more exotic `certfificates_ops::add::add_certificates_batch(certificate_ops)`

what do you think ?

Reply:
Outdated